KEP-1224: Lending Limit to the cohort (#1331)
kubernetes-sigs/kueue: keps/1224-lending-limit/README.md

# KEP-1224: Lending limit

- [Summary](#summary)
- [Motivation](#motivation)
  - [Goals](#goals)
  - [Non-Goals](#non-goals)
- [Proposal](#proposal)
  - [User Stories (Optional)](#user-stories-optional)
    - [Story 1](#story-1)
  - [Risks and Mitigations](#risks-and-mitigations)
- [Design Details](#design-details)
  - [Kueue LendingLimit API](#Kueue LendingLimit API)
  - [Test Plan](#test-plan)
    - [Prerequisite testing updates](#prerequisite-testing-updates)
    - [Unit Tests](#unit-tests)
    - [Integration tests](#integration-tests)
  - [Graduation Criteria](#graduation-criteria)
- [Implementation History](#implementation-history)
- [Drawbacks](#drawbacks)
- [Alternatives](#alternatives)

## Summary

> This proposal provides a guaranteed resource quota for user. By setting `lending limit`, users can have a reservation of resource quota(`nominalQuota` - `lendingLimit`) that will never be borrowed by other clusterqueues in the same cohort.

Reviewer: Let's try to not mention as much technical details as possible, just clarify what we hope to do here, like:
```
In today's Kueue's system, one clusterQueue's resources could be borrowed completely by neighbors inside the same cohort, 
this improves the resource utilization to some extend, but sometimes users want to reserve some resources only for private usage. 
This proposal introduces a new field helps to solve this problem by reserving the demand resources.
```

### Goals

- Implement `LendingLimit`, users can have a reservation of guaranteed resource by claiming the `LendingLimit`.

### Non-Goals

- Replace `BorrowingLimit` to some extent in the future.

## Proposal

In this proposal, `LendingLimit` is defined. The `ClusterQueue` will be limited to lend the specified quota to other ClusterQueues in the same cohort.

### User Stories (Optional)

#### Story 1

In order to ensure the full utilization of resources, we generally set `BorrowingLimit` to max, but this will cause a `ClusterQueue` to run out of its all resources, and make the new incoming job slow to preempt. So we want to reserve some resources for a `ClusterQueue`, so that any incoming jobs in the  `ClusterQueue` can get into work immediately.

### Risks and Mitigations

It's possible that other ClusterQueues in the same cohort may not have enough resources to borrow, because of `LendingLimit`. In other word, `LendingLimit` will influence the resource quota other ClusterQueue's can actually get from `BorrowingLimit`.

## Design Details

### Kueue LendingLimit API

Modify ResourceQuota API object:

```go
type ResourceQuota struct {
	[...]

> 	// lendingLimit is the maximum amount of quota for the [flavor, resource]

Reviewer: This is not solved.

	// lendingLimit must be null if spec.cohort is empty.
	// +optional
	LendingLimit *resource.Quantity `json:"lendingLimit,omitempty"`
}
```

Also `Lended` is added to `ResourceUsage`.

```go
type ResourceUsage struct {
	[...]

    // Lended is quantity of quota that is lended to the cohort. In other
	// words, it's the unused quota that is under the nominalQuota.
    Lended resource.Quantity `json:"lended,omitempty"`
}
```

### Test Plan

[x] I/we understand the owners of the involved components may require updates to
existing tests to make this code solid enough prior to committing the changes necessary
to implement this enhancement.

##### Prerequisite testing updates

None.

#### Unit Tests

- `pkg/cache`: `2023-11-15` - `86.5%`
- `pkg/controller/core/`: `2023-11-15` - `16.5%`
- `pkg/metrics/`: `2023-11-15` - `45.7%`
- `pkg/scheduler/`: `2023-11-15` - `80.6%`

#### Integration tests

<!--
Describe what tests will be added to ensure proper quality of the enhancement.

After the implementation PR is merged, add the names of the tests here.
-->

- No new workloads can be admitted when the `LendingLimit` greater than `NominalQuota` or less than `0`.

> - No new workloads in the cohort can borrow when the sum of `LendingLimit` in the cohort does not satisfy the requirement.

Reviewer: Plz make it more detailed, I think we have at least two test cases here:
- With borrowingLimit unset, one clusterQueue can borrow `totalQuota - sum(LendingLimit)`
- With borrowingLimit set, one clusterQueue can borrow `min((totalQuota - sum(LendingLimit)), borrowingLimit)`

Reviewer: Also, maybe you should add one more test with multiple `lendingLimit` set for different resources.

### Graduation Criteria

## Implementation History

## Drawbacks

## Alternatives

> - `GuaranteedQuota` could do the same. But since we already have the `BorrowingLimit`, it seems more reasonable to use `LendingLimit` that sounds better.

Reviewer: We considered `GuaranteedQuota` as the field name, but `BorrowingLimit` seems more align with the already exist `LendingLimit`,  so we choice the latter.

	// lendingLimit must be null if spec.cohort is empty.
	// +optional
	LendingLimit *resource.Quantity `json:"lendingLimit,omitempty"`
}
```

Also `Lended` is added to `ResourceUsage`.

```go
type ResourceUsage struct {
	[...]

    // Lended is a quantity of quota that is lended to other ClusterQueues in the cohort.
    Lended resource.Quantity `json:"lended,omitempty"`
}
```

### Test Plan

[x] I/we understand the owners of the involved components may require updates to
existing tests to make this code solid enough prior to committing the changes necessary
to implement this enhancement.

##### Prerequisite testing updates

None.

#### Unit Tests

- `pkg/cache`: `2023-11-15` - `86.5%`
- `pkg/controller/core/`: `2023-11-15` - `16.5%`
- `pkg/metrics/`: `2023-11-15` - `45.7%`
- `pkg/scheduler/`: `2023-11-15` - `80.6%`

#### Integration tests

<!--
Describe what tests will be added to ensure proper quality of the enhancement.

After the implementation PR is merged, add the names of the tests here.
-->

- No new workloads can be admitted when the `LendingLimit` greater than `NominalQuota` or less than `0`.
- In a cohort with 2 ClusterQueues a, b:

>   - When cq-b's LendingLimit unset:

Reviewer: I think the test already exists, if not, then add it.

    - When cq-a's BorrowingLimit unset, cq-a can borrow as much as `cq-b's NominalQuota`.
    - When cq-a's BorrowingLimit set, cq-a can borrow as much as `min(cq-b's NominalQuota, cq-a's BorrowingLimit)`.
  - When cq-b's LendingLimit set:
    - When cq-a's BorrowingLimit unset, cq-a can borrow as much as `cq-b's LendingLimit`.

>     - When cq-a's BorrowingLimit set, cq-a can borrow as much as `min(cq-b's LendingLimit, cq-a's BorrowingLimit)`.

Reviewer: Also add a test case which has several resourceFlavors testing the fungibility.

### Graduation Criteria

## Implementation History

## Drawbacks

## Alternatives

> `GuaranteedQuota` could do the same. But since we already have the `BorrowingLimit`, it seems more reasonable to use `LendingLimit` that sounds better.

Reviewer: ```suggestion
`GuaranteedQuota` which defines the quota for reservation is functionally similar to `LendingLimit`, but to align with `BorrowingLimit`, we chose the latter.
```

Even if we set the `Preemption`, it still needs some time and spends a lot of unnecessary cost.

So we need a reservation design for resource requests and security reasons: `LendingLimit`, to claim the quota allowed to lend, reserve a certain amount of resources to ensure that they will never be borrowed.

### Goals

- Implement `LendingLimit`, users can have a reservation of guaranteed resource by claiming the `LendingLimit`.

### Non-Goals

- Replace `BorrowingLimit` to some extent in the future.

## Proposal

In this proposal, `LendingLimit` is defined. The `ClusterQueue` will be limited to lend the specified quota to other ClusterQueues in the same cohort.

### User Stories (Optional)

#### Story 1

In order to ensure the full utilization of resources, we generally set `BorrowingLimit` to max, but this may cause a `ClusterQueue` to run out of its all resources, and make the new incoming job slow to response for the slow preemption. This could be worse in a competitive cluster, jobs will borrow resources and be reclaimed over and over.

So we want to reserve some resources for a `ClusterQueue`, so that any incoming jobs in the  `ClusterQueue` can get admitted immediately.

> ### Risks and Mitigations

Reviewer 2: I think this is more of a Note/Caveat than a risk.

> With both BorrowingLimit and LendingLimit configured, one clusterQueue may not be able to borrow up to the limit just because we reserved the lending limit quota of resource. To aviod this, we can leave the BorrowingLimit unset to borrow as mush as we can.

Reviewer 2: ```suggestion
With both BorrowingLimit and LendingLimit configured, one clusterQueue may not be able to borrow up to the limit just because we reserved the lending limit quota of resource.
To reduce confusion, we will recommend to users to only set borrowingLimit or lendingLimit, but not both, even though both will be supported at the same time.
```

## Design Details

### Kueue LendingLimit API

Modify ResourceQuota API object:

```go
type ResourceQuota struct {
	[...]

> 	// lendingLimit is the maximum amount of quota for the [flavor, resource]
> 	// combination that this ClusterQueue is allowed to lend the unused
> 	// quota to other ClusterQueues in the same cohort.

Reviewer 2: ```suggestion
	// lendingLimit is the maximum amount of unused quota for the [flavor, resource]
	// combination that this ClusterQueue can lend to other ClusterQueues in the same cohort.
```

> 	// In total, at a given time, ClusterQueue can reserve
> 	// a quantity of quota equals to nominalQuota - lendingLimit.

Reviewer 2: ```suggestion
	// In total, at a given time, ClusterQueue reserves for its exclusive use
	// a quantity of quota equals to nominalQuota - lendingLimit.
```

	// If null, it means that there is no lending limit.
	// If not null, it must be non-negative.
	// lendingLimit must be null if spec.cohort is empty.
	// +optional
	LendingLimit *resource.Quantity `json:"lendingLimit,omitempty"`
}
```

Also `Lended` is added to `ResourceUsage`.

```go
type ResourceUsage struct {
	[...]

    // Lended is a quantity of quota that is lended to other ClusterQueues in the cohort.

>     Lended resource.Quantity `json:"lended,omitempty"`

Reviewer 2: I don't think we can safely implement this, because unused resources from multiple CQs compose a single pool of shareable resources.

So there is no concept of A is borrowing from B. A is borrowing from all the unused resource of B, C and any other CQs in the cohort.

I propose we drop this status field.

Reviewer: We can drop this for now, but please add this as a Note for record.

}
```

### Test Plan

[x] I/we understand the owners of the involved components may require updates to
existing tests to make this code solid enough prior to committing the changes necessary
to implement this enhancement.

##### Prerequisite testing updates

None.

#### Unit Tests

- `pkg/cache`: `2023-11-15` - `86.5%`
- `pkg/controller/core/`: `2023-11-15` - `16.5%`
- `pkg/metrics/`: `2023-11-15` - `45.7%`

> - `pkg/scheduler/`: `2023-11-15` - `80.6%`

Reviewer 2: I think the flavorassigner is also relevant.

#### Integration tests

<!--
Describe what tests will be added to ensure proper quality of the enhancement.

After the implementation PR is merged, add the names of the tests here.
-->

- No new workloads can be admitted when the `LendingLimit` greater than `NominalQuota` or less than `0`.
- In a cohort with 2 ClusterQueues a, b and single ResourceFlavor:
  - When cq-b's LendingLimit set:
    - When cq-a's BorrowingLimit unset, cq-a can borrow as much as `cq-b's LendingLimit`.
    - When cq-a's BorrowingLimit set, cq-a can borrow as much as `min(cq-b's LendingLimit, cq-a's BorrowingLimit)`.
- In a ClusterQueue with 2 ResourceFlavors a, b:

>   - When rf-b's LendingLimit set, and FlavorFungibility set to `WhenCanBorrow: Borrow`:

Reviewer 2: this is the default fungibility policy.

Is the feature really impacted by different policies?

Reviewer: Just in case we missed something, this can be intersected with borrowingLimit on exist testcases. I'm ok to miss this case.

    - When rf-b's LendingLimit unset, cq-a can borrow as much as `cq-b's LendingLimit`.

>     - When rf-b's LendingLimit set, cq-a can borrow as much as `min(cq-b's LendingLimit, cq-a's BorrowingLimit)`.

Reviewer 2: It might be worth having a test case with 3 CQs.

### Goals

- Implement `LendingLimit`, users can have a reservation of guaranteed resource by claiming the `LendingLimit`.

### Non-Goals

- Replace `BorrowingLimit` to some extent in the future.

## Proposal

In this proposal, `LendingLimit` is defined. The `ClusterQueue` will be limited to lend the specified quota to other ClusterQueues in the same cohort.

### User Stories (Optional)

#### Story 1

In order to ensure the full utilization of resources, we generally set `BorrowingLimit` to max, but this may cause a `ClusterQueue` to run out of its all resources, and make the new incoming job slow to response for the slow preemption. This could be worse in a competitive cluster, jobs will borrow resources and be reclaimed over and over.

So we want to reserve some resources for a `ClusterQueue`, so that any incoming jobs in the  `ClusterQueue` can get admitted immediately.

### Notes/Constraints/Caveats (Optional)

With both BorrowingLimit and LendingLimit configured, one clusterQueue may not be able to borrow up to the limit just because we reserved the lending limit quota of resource.

> To reduce confusion, we will recommend to users to only set borrowingLimit or lendingLimit, but not both, even though both will be supported at the same time.

Reviewer 3: > I'm ok either way. I remember seeing a similar recommendation in YARN not to use borrowingLimit and weight at the same time.
> 
> Another alternative would be some form of troubleshooting, for example:
> 
> > Why is my CQ not borrowing?
> > Check borrowingLimit of your CQ and lendingLimit of other CQs in the cohort.

I prefer to take this case (`BorrowingLimit` and `LendingLimit`) as a troubleshooting technique.

> To reduce confusion, we will recommend to users to only set borrowingLimit or lendingLimit, but not both, even though both will be supported at the same time.

Author: Sounds reasonable. I agree to remove Line 66.

> To reduce confusion, we will recommend to users to only set borrowingLimit or lendingLimit, but not both, even though both will be supported at the same time.

Reviewer: Plz update the words.

> To reduce confusion, we will recommend to users to only set borrowingLimit or lendingLimit, but not both, even though both will be supported at the same time.

Reviewer 3: > This is not about we should only set `borrowingLimit` or `lendingLimit` or both, but a practice that if we configured the `lendingLimit` for demand, then `borrowingLimit` could be omitted for the maximum resource utilization as there's no borrowingLimit.

@kerthcet Does this mean the following situation?

cq-a and cq-b belong to the same cohort.

- cq-a:
   cpu: 10
   borrowingLimit: 5

- cq-b:
   cpu: 5
   lendingLimit: 0

> To reduce confusion, we will recommend to users to only set borrowingLimit or lendingLimit, but not both, even though both will be supported at the same time.

Reviewer: You can set the borrowingLimit or lendingLimit by your demands, but if we set the lendingLimit in cq-b, then cq-a can leave the borrowingLimit unset to consume the resources as much as possible, as
```
- cq-a:
  cpu: 10
  # borrowingLimit: 5 # no need to set this if you want to maximum the resource utilization 
  # and don't need to worry about use too much resources because other clusterQueues can 
  # constrain this by setting lendingLimit.

- cq-b:
  cpu: 5
  lendingLimit: X
```

### Risks and Mitigations

None.

## Design Details

### Kueue LendingLimit API

Modify ResourceQuota API object:

```go
type ResourceQuota struct {
	[...]

	// lendingLimit is the maximum amount of unused quota for the [flavor, resource]
	// combination that this ClusterQueue can lend to other ClusterQueues in the same cohort.
	// In total, at a given time, ClusterQueue reserves for its exclusive use
	// a quantity of quota equals to nominalQuota - lendingLimit.
	// If null, it means that there is no lending limit.
	// If not null, it must be non-negative.
	// lendingLimit must be null if spec.cohort is empty.
	// +optional
	LendingLimit *resource.Quantity `json:"lendingLimit,omitempty"`
}
```

#### Note

> We have considered adding this status field, but deprecated it in the end. Because unused resources from multiple CQs compose a single pool of shareable resources, and we cannot precisely calculate this value.

Reviewer 2: ```suggestion
We have considered adding this status field, but discarded it. Because unused resources from multiple CQs compose a single pool of shareable resources, we cannot precisely calculate this value.
```
